### ISSUE TITLE
Pyglet on Android and iOS

### ISSUE THREAD

**Issue by QuetzalcoutlDev** [pyglet/pyglet#688]
I would like to make a game in Pyglet but what stops me is that it cannot be for Android, surely many have already mentioned it but I think that current video game development needs to cover all the platforms it can and that Pyglet is available on Android and iOS would be very good

Is there a way to put a project in Pyglet for Android?

**Comment by csd4ni3l** [comment]
pydroid would be the best if it would support it but they dont because pyglet would detect your device as linux and linux in pyglet uses xlib but xlib is really far from android windows because they are not even windows but anyway idk if theres a way but i dont think there is

**Comment by QuetzalcoutlDev** [comment]
> pydroid would be the best if it would support it but they dont because pyglet would detect your device as linux and linux in pyglet uses xlib but xlib is really far from android windows because they are not even windows but anyway idk if theres a way but i dont think there is

If I were an expert in programming, I would look for a way to make Pyglet compatible with Android, but since I am not, I would have to use another module

**Comment by pushfoo** [comment]
> Is there a way to put a project in Pyglet for Android?

Not now, but maybe in the non-immediate future. I think GLES support was added recently, which is a first step toward enabling Android support.

For now, you may want to use [Kivy](https://kivy.org/) if you want to write Android applications using Python. It's more complicated to get started, especially for games, but it has some useful features for mobile.

**Comment by QuetzalcoutlDev** [comment]
> > Is there a way to put a project in Pyglet for Android?
> 
> Not now, but maybe in the non-immediate future. I think GLES support was added recently, which is a first step toward enabling Android support.
> 
> For now, you may want to use [Kivy](https://kivy.org/) if you want to write Android applications using Python. It's more complicated to get started, especially for games, but it has some useful features for mobile.

I think I will use Pygame since we can use it for Android

**Comment by einarf** [comment]
> Is there a way to put a project in Pyglet for Android?

Pyglet is open source and it's for someone passionate about the Android platform to implement that platform support in pyglet. It can probably be done in pyglet 2 (master branch) since it has basic gles support.

**Comment by mingewang** [comment]
any clue/guideline on how to get started?

**Comment by hyuri** [comment]
iOS and Android support would be game-changing. Any updates on this?

Especially now that the [BeeWare](https://github.com/beeware) folks have been getting iOS and Android as officially supported platforms in CPython, and their [Briefcase](https://github.com/beeware/briefcase) tool for deploying to these platforms seems pretty solid.

Would make Pyglet the de-facto and only viable multi-platform Python game development framework.

**Comment by pushfoo** [comment]
*@hyuri please don't spam the same comment on multiple issues.*

**TL;DR:** For now, iOS requires nonexistent tech while Android is a **lot** of work

> [!NOTE]
> This might be partly wrong or out of date. It's a from-memory recap of things wiser minds (einarf, caffeinepills, Ben, and Clepto) have told me.

### General API Changes
The BeeWare tools seem focused around form input / viewing applications, not media. That makes their porting work much easier.

For pyglet, my understanding is there are some API reworks we'd have to do. For example:
* (Likely) Event loop changes to deal with mobile-style pausing and resume
* Both Mac and iOS use doubles or floats for values where pyglet currently uses int (e.g. [mouse position](https://developer.apple.com/documentation/corefoundation/cgpoint))
* We'd need to add touch event support from the ground up
* Media decode needs investigation (ffmpeg isn't viable on iOS, etc)

However, researching these is worthwhile.

### Android is Easier
Android has GLES, which we might currently use on Raspberry Pi.

However, the pre-emption, touch system, and event data type concerns from earlier still apply here.

### No OpenGL on iOS
Even if it turns out we can re-use some of the low-level Objective C bindings and sound bindings:
* there's no OpenGL on iOS
* faking it is *hard*

We'd need an OpenGL <-> Metal bridge which:
1. is feature-complete enough to work
2. has decent performance
3. is licensed under an Apple-friendly license like BSD or MIT (no GPL allowed)

If something like this exists, I am not aware of it. This sort of library is very hard to build since it's going to require in-depth graphics API and transpiler knowledge. That means that if someone writes one like it, it'll likely be one of two things:
* Commercially licensed for profit to help port [Nintendo Switch games built on OpenGL](https://wccftech.com/nintendo-switch-supports-vulkan/)
* GPL-licensed to ensure it can't be sold without benefiting contributors

In either case, it still seems to present issues:
* We'd might need to build a new ctypes binding for it either way
* Even LGPL libraries might not work for app stores since it requires you to be able to ["swap out the DLL"](https://opensource.stackexchange.com/questions/9599/what-is-the-purpose-of-the-lgpl-re-linking-requirement)

**Comment by hyuri** [comment]
@pushfoo Thanks for the update.

I didn't spam though — I deleted the other post and posted here as I figured this would be the better place to post it.

> If something like this exists, I am not aware of it. This sort of library is very hard to build since it's going to require in-depth graphics API and transpiler knowledge. [...]

What about [ANGLE](https://en.wikipedia.org/wiki/ANGLE_(software))? Seems to check all the boxes.

**Comment by pushfoo** [comment]
> What about [ANGLE](https://en.wikipedia.org/wiki/ANGLE_(software))? Seems to check all the boxes.

It's worth investigating further to double check.

My understanding *was* that ANGLE didn't implement the right features yet. This is based on:
* past discussion with @Cleptomania in the Arcade Discord
* [this table on the Wikipedia page](https://en.wikipedia.org/wiki/ANGLE_(software)#Level_of_OpenGL_ES_support_via_backing_renderers)

Note that I might be misreading it and that it may be out of date.

*@mingewang since you asked how to get started earlier in the thread, I'm tagging you too.*

If anyone can get ANGLE working, it'd go a long way toward drumming up enthusiasm for mobile. There's currently a chicken-and-egg situation. Touch events, suspend/resume, and graphics API problems look like a giant unsolvable monolith which discourages people from touching any part of it.

### Before starting

It'd be worth looking at:
1. The data types for iOS and Android touch and screen positions
   * Ideally, we'd build a unified model of cross-platform touch
   * Please don't repeat the strange import tricks currently used for `pyglet.Window` ; - ;
   * `typing.Protocol` helps
   * Be aware that mypy doesn't like `NamedTuple`
3. How Kivy and Beeware handle their suspend / resume (if the licenses are compatible!)
4. The pyglet event loop
5. The Objective C bindings

#### How pyglet binds Objective C

First, there's some Objective C binding code in the [`pyglet/lib/darwin`](https://github.com/pyglet/pyglet/tree/master/pyglet/libs/darwin) folder:
1. It isn't documented the same way since it's internal-only (this applies to all of `pyglet/lib`)
2. At least some of it is vendored libraries others and I have made some updates to for compat / clarity
3. It's still very cursed (many comments like "I don't know?" and "Maybe?")


### First Steps for MVP
For a rendering-only proof of concept, my understanding is we'd need to start with the following:
1. Figure out which features we currently use to get GLES working (whatever's used for RasPi)
2. Demonstrate GLES <-> Metal works on the required subset on iOS
3. Get pyglet itself rendering on iOS, though possibly with input handling stubbed / disabled on a branch
4. A prototype ctypes binding with wake / resume support

Note that  pyglet has a hard 0-dependency rule. That means event / resume would have to be new or vendored from a library with a compatible license instead of an external import.

### Post-MVP

My understanding is that the Mac code overlaps with the iOS due to sharing Objective C / cocoa infrastructure. Maybe only a few things, but...

#### Refactor Cocoa / Implement Touch

The Cocoa binding is better than the base Objective C one. It lives at [`pyglet/gl/cocoa.py`](https://github.com/pyglet/pyglet/blob/master/pyglet/gl/cocoa.py). Note that Cocoa and [Cocoa Touch](https://developer.apple.com/library/archive/documentation/General/Conceptual/DevPedia-CocoaCore/Cocoa.html) are technically separate frameworks, but maybe there's some overlap?

I can't really help you here since I don't have a working Mac or Apple device at the moment.

#### Finish The GL Back-End?
I'm not sure how to do this beyond:
1. Tests
2. Type annotations
3. Doc
Although I've gone deep into our shader stack once, it was a while ago and to work around a problem in tests. Ben ended up solving it better in another way.

**Comment by Cleptomania** [comment]
As far as the GL version requirements are concerned, as far as I’m aware Pyglet should work fine with GL ES 3.1, which ANGLE seems to support.

I know that Arcade is fully functional with 3.1 + Geometry shader extension, and Pyglet doesn’t use geometry shaders out of the box, so I don’t see why Pyglet shouldn’t work with only 3.1 features. Technically geometry shaders weren’t core in ES until 3.2, but many 3.1 drivers expose the extension for them.
